tests/kernel/common/src/rand32.c

```diff
@@ -43,12 +43,15 @@ void rand32_test(void)
 	 * and save results in an array to verify them on the
 	 * next step
 	 */
-	for (i = 0; i < N_VALUES; i++) {
-		rnd_values[i] = sys_rand32_get();
-	}
-	for (i = 1; i <  N_VALUES; i++) {
-		zassert_false((rnd_values[i - 1] == rnd_values[i]),
-			      "random number subsequent calls return same value");
+	for (rnd_cnt = 0; rnd_cnt < N_VALUES; rnd_cnt++) {
+		rnd_values[rnd_cnt] = sys_rand32_get();
 	}
 
+	for (rnd_cnt = 0; rnd_cnt <  N_VALUES; rnd_cnt++) {
+		for (cmp_cnt = rnd_cnt+1; cmp_cnt <  N_VALUES; cmp_cnt++) {
+			zassert_false(
+				(rnd_values[rnd_cnt] == rnd_values[cmp_cnt]),
```

Comment:
I don't think there's improvement over the original code. There's nothing wrong with RNG returning the same number twice. (It's just probability of that should be low.)

Reply:
agreed, this can fail even if everything is working properly.
@lpereira any opinion?

Reply:
I don't know what's the purpose of this patch, either.

There are no guarantees on how the random numbers generated by sys_rand32_get() will be distributed.  Ideally, this test should test only the API, without testing the random number generator itself -- difficult to do, without knowing what to look for. Even when not testing the random number themselves, this is hard to do by doing a black box test.

Having said this: this test should indeed see if sys_rand32_get() returned the same number twice, but it shouldn't fail immediately if it returns the same number.  Something like this:

```
equal_count := 0
generated_numbers := 0
last_gen := sys_rand32_get()

for i := 0 to N_VALUES-1 {
      gen := sys_rand32_get()
      if gen != last_gen {
          equal_count++
      }
      last_gen := gen
}
if equal_count > N_VALUES / 2 {
      fail_test();
} else {
      pass_test();
}
```

This will only test if successive calls returned the same number, and only fail if half of the numbers generated were the same as the previously generated one. If the first generated number is the same as the last one (for instance), this won't be an issue.

This should catch a buggy sys_rand32_get() function, or cases where rand32_entropy_device is used and the entropy driver fails to signal error when no entropy is available.

Any comments?  I don't know how else to test this, to be honest.

Reply:
Thank you for the information. We were not able to find a clear documentation on sys_rand32_get() function. It is helpful to understand the function.
So, we were checking for each random numbers to be unique. Code is updated as per suggestion now.